janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java

```diff
@@ -173,12 +370,59 @@ public boolean isValidCondition(Object condition) {
             return condition instanceof String && StringUtils.isNotBlank(condition.toString());
         }
 
+        @Override
+        public boolean hasNegation() {
+            return true;
+        }
+
+        @Override
+        public JanusGraphPredicate negate() {
+            return Text.NOT_REGEX;
+        }
+
         @Override
         public String toString() {
             return "textRegex";
         }
 
-    }, 
+    },
+
+    /**
+     * Whether the text fails a regular expression (case sensitive)
+     */
+    NOT_REGEX {
+
+        @Override
+        public boolean test(Object value, Object condition) {
+            this.preevaluate(value, condition);
+            return value != null && evaluateRaw(value.toString(), (String) condition);
```

Comment:
@porunov / @li-boxuan / @farodin91 

Sorry to bring you all back to this PR, but after using this some more locally, I ran into a case that was not covered here regarding null values / missing properties:

```
gremlin> g.V().has('missingField', textNotRegex('.*'))
==>v[86040]
gremlin> g.V(86040L).propertyMap()
==>[]
gremlin> g.V(86040L).has('missingField', textNotRegex('.*'))
gremlin>
```

Should the negation tests remove the != null tests, or just simply short circuit if null?

```
return value == null || evaluateRaw(value.toString(), (String) condition);
```

Reply:
I went ahead and created a new PR for this change (https://github.com/JanusGraph/janusgraph/pull/2593). If you dont think `.has(X, Y)` assumes `.has(X).has(X, Y)`, feel free to close it.

Reply:
> gremlin> g.V().has('missingField', textNotRegex('.*'))
==>v[86040]

I believe above is wrong. I actually suspect this will fail for TinkerPop test suite, since I don't believe this is TinkerPop-aligned. A related discussion is on https://github.com/JanusGraph/janusgraph/issues/1868

My understanding is:
`g.V().has('missingField', textNotRegex('.*')) == g.V().has('missingField', not(textRegex('.*'))) == g.V().has('missingField').has('missingField', not(textRegex('.*'))) != g.V().not(has('missingField', textRegex('.*')))`

Reply:
Perfect. I updated my new PR to change the negated query in ES from `mustNot(predicate)` to `must(exists(field), mustNot(predicate))`, which aligns the predicate evaluations:

```
gremlin> g.V().has('missingField', textNotRegex('.*'))
gremlin>
gremlin> g.V(86040L).propertyMap()
==>[]
gremlin> g.V(86040L).has('missingField', textNotRegex('.*'))
gremlin>
```